sim/midas/src/main/scala/midas/passes/fame/PromotePassthroughConnections.scala

```diff
@@ -0,0 +1,136 @@
+  // scheduling of these passes in a different order. Define these for
+  // future-proofing & documentation.
+	override def prerequisites = Forms.LowForm :+ Dependency[ExtractModel]
+  override def optionalPrerequisiteOf = Seq(Dependency[FAMEDefaults])
+
+  def execute(state: CircuitState): CircuitState = {
+    val moduleMap = state.circuit.modules.map(m => m.name -> m).toMap
+    val topModule = moduleMap(state.circuit.main)
+
+    def collectInstances(insts: mutable.ArrayBuffer[WDefInstance])(s: Statement): Unit = {
+      s match {
+        case wdef: WDefInstance => insts += wdef
+        case o => Nil
+      }
+      s.foreach(collectInstances(insts))
+    }
+
+    val modelInstances = new mutable.ArrayBuffer[WDefInstance]()
+    topModule.foreach(collectInstances(modelInstances))
+    val instanceNames = modelInstances.map(_.name).toSet
+
+    val modelNodes = modelInstances.flatMap { case WDefInstance(_, instName, modName, _) =>
+      moduleMap(modName).ports
+        .filterNot(_.tpe == ClockType)
+        .map(p => p.direction -> LogicNode(p.name, inst = Some(instName)))
+    }
+    val modelSources = modelNodes.collect { case (Output, ln) => ln }
+    val modelSinks   = modelNodes.collect { case (Input, ln) => ln }
+
+    val bridgeSources = topModule.ports.collect {
+      case Port(_, name, Input, tpe) if tpe != ClockType => LogicNode(name)
+    }
+    val bridgeSinks   = topModule.ports.collect {
+      case Port(_, name, Output, tpe) if tpe != ClockType => LogicNode(name)
+    }
+
+    val allSinks = modelSinks ++ bridgeSinks
+    val allSources = modelSources ++ bridgeSources
+
+    val connectivity = new CheckCombLoops().analyzeFull(RemoveNonWirePrimitives(state))(state.circuit.main)
+    val originalSources = allSources.filter { s => connectivity.reachableFrom(s).isEmpty }.toSet
+
+    val sink2sourceMap = (for (sink <- allSinks) yield {
+      val source = connectivity.reachableFrom(sink) & originalSources
+      assert(source.size == 1)
+      sink -> source.head
+    }).toMap
+
+
+    def onStmt(s: Statement): Statement = s.map(onStmt) match {
+      case c@Connect(_, lhs, _) =>
+        val sinkNode = lhs match {
+          case WSubField(WRef(instName,_,InstanceKind,_), portName, tpe, _) if tpe != ClockType =>
+            Some(LogicNode(portName, Some(instName)))
+          case WRef(name, tpe, _, _) if tpe != ClockType => Some(LogicNode(name))
+          case o => None
+        }
+        sinkNode match {
+          case Some(sinkNode) =>
+            val sourceNode = sink2sourceMap(sinkNode)
+            val newRHS = sourceNode match {
+              case LogicNode(portName, Some(instName), None) => WSubField(WRef(instName), portName)
+              case LogicNode(portName, None, None) => WRef(portName)
+              case o => throw new Exception(s"memport field of source LogicNode should be unset.") 
+            }
+            c.copy(expr = newRHS)
```

Comment:
This approach (anything driven by a passthrough gets directly driven by the upstream source) is reasonable. Since it happens before channel excision, it also looks like we could use DCE to delete any newly-unused wires.